### PR TITLE
Cfq fix metrics

### PR DESCRIPTION
**Pull request by Ranapop**
The changes mostly modify accuracy computation so that the sequences of predictions and labels are compared directly, without being transformed to a one hot encoding first. The changes were done mainly to avoid unnecessary computation, but also make the behaviour more clear. 

Before these changes, during evaluation, if the predictions needed to be padded, they would have been padded with a zero-valued vector of vocabulary size- which is not a valid ohe vector. In the numpy argmax implementation, the first index would be returned for such a vector, coincidently the index of the padding element, but I would say it's best not to rely on such unintuitive behaviour of used functions for correct implementation.

Moreover, in these changes the padding function necessary for metrics computation is changed to a more general function which takes as input the axis to do the padding on. The function now uses jnp (numpy from jax) but could be easily modified to numpy if needed.

These changes were tested on both the dummy and cfq data and led to no visible change in loss and accuracy values.

### CODE REVIEW

**Review by marcvanzee — CHANGES_REQUESTED** [review]
As a general comment: Could you please explain the PR description what this change does? Did you test it? What was the result?

File: examples/cfq/train.py [hunk 1]
```diff
@@ -80,56 +80,59 @@ def create_model(vocab_size: int) -> nn.Module:
 
 
 def cross_entropy_loss(logits: jnp.array, labels: jnp.array,
-                       lengths: jnp.array):
+                       lengths: jnp.array, vocab_size: int):
   """Returns cross-entropy loss."""
+  labels = common_utils.onehot(labels, vocab_size)
```
Comment (marcvanzee) [inline examples/cfq/train.py:85]: Ah I see, so you moved the OHE into this loss function, so you can remove it in the other places. Nice!

File: examples/cfq/train.py [hunk 2]
```diff
@@ -80,56 +80,59 @@ def create_model(vocab_size: int) -> nn.Module:
...
-      seq_len: 2nd dimension current value
-      max_len: 2nd dimension desired value
-    """
-  padding_size = max_len - seq_len
-  padding = tf.constant([[0, 0], [0, padding_size], [0, 0]])
-  return tf.pad(batch, padding, "CONSTANT").numpy()
+def pad_along_axis(array: jnp.array,
+                   padding_size: int,
+                   axis: int) -> jnp.array:
+  """Returns array padded on axis to target_len"""
```
Comment (marcvanzee) [inline examples/cfq/train.py]: What is `target_len`? Should this be `padding_size`?

File: examples/cfq/train.py [hunk 3]
```diff
@@ -80,56 +80,59 @@ def create_model(vocab_size: int) -> nn.Module:
...
+  ndim = array.ndim
+  pad_shape = jnp.full((ndim, 2), 0)
```
Comment (marcvanzee) [inline examples/cfq/train.py]:
```suggestion
  pad_shape = jnp.zeros((array.ndim, 2))
```

Comment (marcvanzee) [inline examples/cfq/train.py]: Why do we have to create a matrix of dimension `(array.ndim, 2)`? What is the `2` for?

File: examples/cfq/train.py [hunk 4]
```diff
@@ -80,56 +80,59 @@ def create_model(vocab_size: int) -> nn.Module:
...
+  pad_shape = jax.ops.index_update(pad_shape,
+                                    jax.ops.index[axis,1],
+                                    padding_size)
+  padded = jnp.pad(array,pad_shape)
```
Comment (marcvanzee) [inline examples/cfq/train.py]:
```suggestion
  pad_shape = jax.ops.index_update(pad_shape,
                                   jax.ops.index[axis, 1],
                                   padding_size)
  padded = jnp.pad(array, pad_shape)
```

**Review by marcvanzee — CHANGES_REQUESTED** [review]
Thanks for the extensive explanation in the PR description, very helpful!

File: examples/cfq/train.py (hunk 2 above)
Comment (Ranapop) [reply examples/cfq/train.py]: No, target_len is the desired length of the output. I will reformulate

File: examples/cfq/train.py (hunk 3 above)
Comment (Ranapop) [reply examples/cfq/train.py]: The pad function expects an array specifying the padding on each axis, at the begining and end. So it expects something like [[before_0,after_0],...[before_n,after_n]] where n is the number of dimensions. As you can see, for each dimension there's a vector of dimension 2 (before, after). I'm not sure how to put it in a short comment or how else to improve the code readability here. I will try to add something

File: examples/cfq/train.py [hunk 5]
```diff
@@ -80,56 +80,62 @@ def create_model(vocab_size: int) -> nn.Module:
...
-  padding = tf.constant([[0, 0], [0, padding_size], [0, 0]])
-  return tf.pad(batch, padding, "CONSTANT").numpy()
+def pad_along_axis(array: jnp.array,
+                   padding_size: int,
+                   axis: int) -> jnp.array:
+  """Returns array padded with zeroes on given axis with padding_size positions.
+  The padding is done at the end of the array at that axis"""
+  # The pad function expects a shape of size (n,2) where n is the number of axes
+  # For each axis, the number of padding positions at the begining and end of
+  # the array is specified
```
Comment (marcvanzee) [inline examples/cfq/train.py]:
```suggestion
  # The pad function expects a shape of size (n, 2) where n is the number of axes.
  # For each axis, the number of padding positions at the beginning and end of
  # the array is specified.
```

File: examples/cfq/train.py [hunk 6]
```diff
@@ -80,56 +80,62 @@ def create_model(vocab_size: int) -> nn.Module:
...
-      max_len: 2nd dimension desired value
-    """
-  padding_size = max_len - seq_len
-  padding = tf.constant([[0, 0], [0, padding_size], [0, 0]])
-  return tf.pad(batch, padding, "CONSTANT").numpy()
+def pad_along_axis(array: jnp.array,
+                   padding_size: int,
+                   axis: int) -> jnp.array:
+  """Returns array padded with zeroes on given axis with padding_size positions.
+  The padding is done at the end of the array at that axis"""
```
Comment (marcvanzee) [inline examples/cfq/train.py]:
```suggestion
  The padding is done at the end of the array at that axis."""
```

File: examples/cfq/train.py [hunk 7]
```diff
@@ -80,56 +80,62 @@ def create_model(vocab_size: int) -> nn.Module:
...
+  # The pad function expects a shape of size (n,2) where n is the number of axes
+  # For each axis, the number of padding positions at the begining and end of
+  # the array is specified
+  pad_shape = jnp.full((array.ndim, 2), 0)
```
Comment (marcvanzee) [inline examples/cfq/train.py]: Is there are reason you are not using `jnp.zeros` here, like I suggested?

```suggestion
  pad_shape = jnp.zeros((array.ndim, 2))
```

**Review by marcvanzee — APPROVED** [review]

File: examples/cfq/train.py (hunk 7 above)
Comment (Ranapop) [reply examples/cfq/train.py]: no
